homeassistant/components/zha/sensor.py

```diff
@@ -271,18 +149,125 @@ def state(self) -> str:
-        self._state = self._formatter_function(state)
+        if state is not None:
+            state = self.formatter(state)
+        self._state = state
         self.async_schedule_update_ha_state()
 
     @callback
     def async_restore_last_state(self, last_state):
         """Restore previous state."""
         self._state = last_state.state
         self._unit = last_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+
+    @callback
+    async def async_state_attr_provider(self):
+        """Initialize device state attributes."""
+        return {}
+
+    def formatter(self, value):
+        """Numeric pass-through formatter."""
+        if self._decimals > 0:
+            return round(
+                float(value * self._multiplier) / self._divisor, self._decimals
+            )
+        return round(float(value * self._multiplier) / self._divisor)
+
+
+@ZHA_ENTITIES.strict_match(MatchRule(channel_names={CHANNEL_POWER_CONFIGURATION}))
+class Battery(Sensor):
+    """Battery sensor of power configuration cluster."""
+
+    _device_class = DEVICE_CLASS_BATTERY
+    _unit = "%"
+
+    @staticmethod
+    def formatter(value):
+        """Return the state of the entity."""
+        # per zcl specs battery percent is reported at 200% ¯\_(ツ)_/¯
+        if not isinstance(value, numbers.Number) or value == -1:
+            return value
+        value = round(value / 2)
+        return value
+
+    async def async_state_attr_provider(self):
+        """Return device statr attrs for battery sensors."""
+        state_attrs = {}
+        battery_size = await self._channel.get_attribute_value("battery_size")
+        if battery_size is not None:
+            state_attrs["battery_size"] = BATTERY_SIZES.get(battery_size, "Unknown")
+        battery_quantity = await self._channel.get_attribute_value("battery_quantity")
+        if battery_quantity is not None:
+            state_attrs["battery_quantity"] = battery_quantity
+        return state_attrs
+
+
+@ZHA_ENTITIES.strict_match(MatchRule(channel_names={CHANNEL_ELECTRICAL_MEASUREMENT}))
+class ElectricalMeasurement(Sensor):
+    """Active power measurement."""
+
+    _device_class = DEVICE_CLASS_POWER
+    _divisor = 10
+    _unit = POWER_WATT
+
+    @property
+    def should_poll(self) -> bool:
+        """Return True if HA needs to poll for state changes."""
+        return True
+
+
+@ZHA_ENTITIES.strict_match(MatchRule(generic_ids={CHANNEL_ST_HUMIDITY_CLUSTER}))
+@ZHA_ENTITIES.strict_match(MatchRule(channel_names={CHANNEL_HUMIDITY}))
+class Humidity(Sensor):
+    """Humidity sensor."""
+
+    _device_class = DEVICE_CLASS_HUMIDITY
+    _divisor = 100
+    _unit = "%"
+
+
+@ZHA_ENTITIES.strict_match(MatchRule(channel_names={CHANNEL_ILLUMINANCE}))
+class Illuminance(Sensor):
+    """Illuminance Sensor."""
+
+    _device_class = DEVICE_CLASS_ILLUMINANCE
+    _unit = "lx"
+
+    @staticmethod
+    def formatter(value):
+        """Convert Illimination data."""
```

Comment:
Convert illumination data.

Reply:
and I just copied it :) Good catch, fixed in https://github.com/home-assistant/home-assistant/pull/30108/commits/df2e1f32fb62a0a11ab33fd7e996c109ecfcd160